pyccl/halos/concentration.py

```diff
@@ -0,0 +1,308 @@
+        none is passed at initialization.
+        """
+        self.mdef = MassDef('fof', 'matter')
+
+    def _setup(self):
+        """ Use this function to initialize any internal attributes
+        of this object. This function is called at the very end of the
+        constructor call.
+        """
+        pass
+
+    def _check_mdef(self, mdef):
+        """ Return False if the input mass definition agrees with
+        the definitions for which this concentration-mass relation
+        works. True otherwise. This function gets called at the
+        start of the constructor call.
+
+        Args:
+            mdef (:obj:`MassDef`): a mass definition object.
+
+        Returns:
+            bool: True if the mass definition is not compatible with
+                this parametrization. False otherwise.
+        """
+        return False
+
+    def _get_consistent_mass(self, cosmo, M, a, mdef_other):
+        """ Transform a halo mass with a given mass definition into
+        the corresponding mass definition that was used to initialize
+        this object.
+
+        Args:
+            cosmo (:obj:`Cosmology`): A Cosmology object.
+            M (float or array_like): halo mass in units of M_sun.
+            a (float): scale factor.
+            mdef_other (:obj:`MassDef`): a mass definition object.
+
+        Returns:
+            float or array_like: mass according to this object's
+            mass definition.
+        """
+        if mdef_other is not None:
+            M_use = mdef_other.translate_mass(cosmo, M, a, self.mdef)
+        else:
+            M_use = M
+        return M_use
+
+    def get_concentration(self, cosmo, M, a, mdef_other=None):
+        """ Returns the concentration for input parameters.
+
+        Args:
+            cosmo (:obj:`Cosmology`): A Cosmology object.
+            M (float or array_like): halo mass in units of M_sun.
+            a (float): scale factor.
+            mdef_other (:obj:`MassDef`): the mass definition object
+                that defines M.
+
+        Returns:
+            float or array_like: concentration.
+        """
+        M_use = self._get_consistent_mass(cosmo,
+                                          np.atleast_1d(M),
+                                          a, mdef_other)
+
+        c = self.concentration(cosmo, M_use, a)
+        if np.isscalar(M):
+            c = c[0]
+        return c
+
+
+class ConcentrationDiemer15(Concentration):
+    name = 'Diemer15'
+
+    def __init__(self, mdef=None):
+        super(ConcentrationDiemer15, self).__init__(mdef)
+
+    def _default_mdef(self):
+        self.mdef = MassDef(200, 'critical')
+
+    def _setup(self):
+        self.kappa = 1.0
+        self.phi_0 = 6.58
+        self.phi_1 = 1.27
+        self.eta_0 = 7.28
+        self.eta_1 = 1.56
+        self.alpha = 1.08
+        self.beta = 1.77
+
+    def _check_mdef(self, mdef):
+        if (np.fabs(mdef.Delta - 200.) > 1E-4) and \
```

Comment:
I tend to agree with the comments we got that this is not safe. I expect the failures are rare, but I feel like we should have a better solution

Reply:
The instances of this we got comments about where things like `if Delta != 200` or even worse `if Delta != ccl.delta_vir(blah)`, which made some of our mass functions completely unusable. I think if someone passes 200.0001 or 199.9999, they actually meant 200. I did set up specific Delta values as strings for `fof` and `vir`. I could set up one for `200`, since it's a common one...

Reply:
We could force casts to int to solve this.

Reply:
We could force casts to int to solve this.

Reply:
so like `if int(mdef.Delta) != 200`?

Reply:
yeah we could do it like this. Better yet to force all delta to be ints on the mdef object.

Reply:
alas, but there are mass functions that work for a continuous range of Deltas...
If you're OK with checking for ints in the specific cases of 200 and 500, I'll implement that.

Reply:
This is probably fine.

Reply:
OK, done.